_posts/2024-03-13-bitcoin-inquisition-#45.md

```diff
@@ -11,17 +11,109 @@ status: upcoming
+off chain UTXO sharing protocols and more. Concretely, it enables LN-Symmetry,
+Timeout Trees, simplified PTLC scripts, unidirectional non-interactive
+channels, (better than now) vaults, trustless coin pools, and more.
+
+
+
+As Inquisition already contains an implementation of BIP-119
+(`OP_CHECKTEMPLATEVERIFY`), implementing LNHANCE on Inquisition only requires
+adding implementations of BIN-2024-0003 and -0004.
+
+This Pull Request adds a total of 3 opcodes:
+* `OP_INTERNALKEY` in Tapscript only, copies the Taproot Internal Key to the
+  stack.
+* `OP_CHECKSIGFROMSTACK` in Tapscript only, validates a BIP340 Schnorr
+  signature using an item from the stack as the message.
+* `OP_CHECKSIGFROMSTACKVERIFY` in all script types, validates a BIP340 Schnorr
+  signature using an item from the stack as the message, and leaves the stack
+  unchanged.
+
+The semantics of the signature-checking opcodes are similar to Tapscript
+opcodes, including the empty signature->false behavior for
+`OP_CHECKSIGFROMSTACK`.
+
+### [The code](https://github.com/bitcoin-core-review-club/bitcoin/tree/prbitcoin-inquisition-45)
+
+#### `OP_INTERNALKEY`
+
+Because this PR is for bitcoin inquisition, the
+[implementation](https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/script/interpreter.cpp#L1309)
+of `OP_INTERNALKEY` is remarkably simple. The taproot internal key is already
+extracted for use with verifying tapscript `OP_CHECKSIG(VERIFY|ADD)` with the
+key `0x01`, so it simply needs to be added to the stack.
+
+#### `OP_CHECKSIGFROMSTACK(VERIFY)`
+
+The implementation of CSFS(V) is split between the [main opcode
+processing](https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/script/interpreter.cpp#L1318), and a function analogous to `EvalChecksigPreTapscript` and `EvalChecksigTapscript` named [`EvalChecksigFromStack`](https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/script/interpreter.cpp#L346).
 
+An earlier approach was to add a `SignatureChecker` that could carry the
+message rather than generating the message from the transaction, and give CSFS
+the exact semantics of existing sigops in their respective script types, but
+that approach had two problems: getting BIP340 signing into legacy/segwitv0
+scripts became intrusive, and the semantics of legacy/segwitv0 signature
+checking are not particularly desirable for a new opcode. As a result the
+current implementation uses a separate function specific to CSFS and gives
+modern Tapscript-style semantics. Only after changing code approaches did it
+become clear that ECDSA checking might not be desirable in these new sigops.
+The commit history still reflects the removal of ECDSA in a separate commit.
+
+Outside of these changes in `interpreter.cpp`, the only non-flag change in
+production code is to allow variable length input to BIP340
```

Comment:
Confused as to what you mean by "production code." Also I don't think the term "flag change" (and "non-flag change") is defined?